File: 108.Convert-Sorted-Array-to-Binary-Search-Tree.py
# https://leetcode.com/problems/convert-sorted-array-to-binary-search-tree/
#
# algorithms
# Easy (48.15%)
# Total Accepted:    222,749
# Total Submissions: 462,569

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def sortedArrayToBST(self, nums):
        """
        :type nums: List[int]
        :rtype: TreeNode
        """
        length = len(nums)
        
        if length == 0:
            return None
        
        if length == 1:
            return TreeNode(nums[0])
        
        middle = length / 2
        node = TreeNode(nums[middle])
        node.left = self.sortedArrayToBST(nums[:middle])
        node.right = self.sortedArrayToBST(nums[middle + 1:])
        
        return node

File: 108.Convert-Sorted-Array-to-Binary-Search-Tree.go
// https://leetcode.com/problems/convert-sorted-array-to-binary-search-tree/
//
// algorithms
// Easy (52.52%)
// Total Accepted:    294,117
// Total Submissions: 560,015

package leetcode

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func sortedArrayToBST(nums []int) *TreeNode {
	length := len(nums)
	if length == 0 {
		return nil
	}
	if length == 1 {
		return &TreeNode{
			Val: nums[0],
		}
	}

	middle := length / 2
	node := &TreeNode{
		Val: nums[middle],
	}
	node.Left = sortedArrayToBST(nums[:middle])
	node.Right = sortedArrayToBST(nums[middle+1:])

	return node
}
